## ECBM E6040 Homework 4 - Programming Problem

In [ ]:
import matplotlib
import pylab
%matplotlib inline

### Problem a: Recurrent Neural Network
#### Bullet 1: Experiment with RNNSLU

In [ ]:
# TODO
from hw4a import test_rnnslu
import gc
question1Output = []
gc.collect()
for nhidden_units in [100,200]:
    for window_size in [7,15]:
        for emb_dim in [50,60]:
            print "Number of hidden units: %s, Context window size: %s, Word embedding size: %s"%(nhidden_units,window_size,emb_dim)
            question1Output.append(test_rnnslu(fold=3, lr=0.0970806646812754, verbose=True, decay=True, win=window_size, 
                                          nhidden=nhidden_units, seed=345, emb_dimension=emb_dim, nepochs=60))
            gc.collect()
print question1Output

Number of hidden units: 100, Context window size: 7, Word embedding size: 50
verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model

Using gpu device 0: GRID K520
WARNING (theano.gof.compilelock): Overriding existing lock by dead process '4517' (I am process '5896')



... training
NEW BEST: epoch 0 valid F1 63.39 best test F1 60.51
NEW BEST: epoch 1 valid F1 75.46 best test F1 72.37
NEW BEST: epoch 2 valid F1 80.34 best test F1 77.06
NEW BEST: epoch 3 valid F1 83.89 best test F1 80.94
NEW BEST: epoch 4 valid F1 85.89 best test F1 82.83
NEW BEST: epoch 5 valid F1 88.02 best test F1 84.35
NEW BEST: epoch 6 valid F1 89.36 best test F1 85.14
NEW BEST: epoch 7 valid F1 90.96 best test F1 87.83
NEW BEST: epoch 8 valid F1 91.66 best test F1 88.68
NEW BEST: epoch 9 valid F1 93.03 best test F1 89.86
NEW BEST: epoch 10 valid F1 93.53 best test F1 90.73
NEW BEST: epoch 11 valid F1 94.0 best test F1 91.15
NEW BEST: epoch 12 valid F1 94.31 best test F1 91.34
NEW BEST: epoch 13 valid F1 94.62 best test F1 92.11
NEW BEST: epoch 14 valid F1 94.74 best test F1 91.72
NEW BEST: epoch 15 valid F1 95.54 best test F1 92.5
NEW BEST: epoch 16 valid F1 95.61 best test F1 92.79

NEW BEST: epoch 18 valid F1 95.93 best test F1 93.05
NEW BEST: epoch 19 valid F1 96.07 best test

### a.i Experiment with RNNSLU

Testing accuracy achieved is stated below:
```
[hidden units, context window, embedding dimension -------- Testing accuracy]
  100,             7,           50                            93.89
  100,             7,           60                            93.61
  100,            15,           50                            93.64
  100,            15,           60                            93.92
  200,             7,           50                            93.92
  200,            15,           60                            94.07
  200,            15,           50                            93.66
  200,            15,           60                            93.91
```

The testing accuracy consistently stays within range of 93.89 +/- 0.6 percent. However, we can see that when the embedding dimension increases the testing accuracy decreases.

#### Bullet 2: Experiment with RNNSLU without normalizing word embeddings

In [1]:
# TODO
from hw4a import test_rnnslu
import gc
question2Output = []
gc.collect()
for nhidden_units in [100,200]:
    for window_size in [7,15]:
        for emb_dim in [50,60]:
            print "Number of hidden units: %s, Context window size: %s, Word embedding size: %s"%(nhidden_units,window_size,emb_dim)
            tmp = test_rnnslu(fold=3, lr=0.0970806646812754, verbose=True, decay=True, win=window_size, 
                                          nhidden=nhidden_units, seed=345, emb_dimension=emb_dim, nepochs=60,normal=False)
            print "Iteration BEST Test: ",tmp
            question2Output.append(tmp)
            gc.collect()
print question2Output

Number of hidden units: 100, Context window size: 7, Word embedding size: 50
verbose: True
normal: False
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 63.39 best test F1 60.51
NEW BEST: epoch 1 valid F1 75.46 best test F1 72.37
NEW BEST: epoch 2 valid F1 80.34 best test F1 77.06
NEW BEST: epoch 3 valid F1 83.89 best test F1 80.94
NEW BEST: epoch 4 valid F1 85.89 best test F1 82.83
NEW BEST: epoch 5 valid F1 88.02 best test F1 84.35
NEW BEST: epoch 6 valid F1 89.36 best test F1 85.14
NEW BEST: epoch 7 valid F1 90.96 best test F1 87.83
NEW BEST: epoch 8 valid F1 91.66 best test F1 88.68
NEW BEST: epoch 9 valid F1 93.03 best test F1 89.86
NEW BEST: epoch 10 valid F1 93.53 best test F1 90.73
NEW BEST: epoch 11 valid F1 94.0 best test F1 91.15
NEW BEST: epoch 12 valid F1 94.31 best test F1 91.34
NEW BEST: epoch 13 val

Using gpu device 0: GRID K520


### a.ii Experiment with RNNSLU without normalizing word embeddings
Testing accuracy across 8 iterations with normal = False is as follows:
```
[hidden units, context window, embedding dimension -------- Testing accuracy]
  100,             7,           50                            93.89
  100,             7,           60                            93.61
  100,            15,           50                            93.64
  100,            15,           60                            93.93
  200,             7,           50                            93.92
  200,             7,           60                            93.99
  200,            15,           50                            93.66
  200,            15,           60                            93.91
```

When the embeddings are not normalized, the testing accuracy decreases. However, the testing accuracy increases as the number of hidden units increase. 

#### Bullet 3: Implement RNN with two hidden layers

In [ ]:
# TODO
from hw4a import test_rnnslu2
import gc
question3Output = []
gc.collect()
for nhidden1 in [100,200]:
    for nhidden2 in [100,200]:
        for window_size in [7,15]:
            for emb_dim in [50,60]:
                print "Number of hidden units 1: %s, Number of hidden units 2: %s, Context window size: %s, Word embedding size: %s"%(nhidden1,nhidden2,window_size,emb_dim)
                tmp = test_rnnslu2(fold=3, lr=0.0970806646812754, verbose=True, decay=True, win=window_size,
                                  nhidden1=nhidden1,nhidden2=nhidden2, seed=345, emb_dimension=emb_dim,
                                  nepochs=20)
                print "Iteration BEST test score: ",tmp
                question3Output.append(tmp)
                gc.collect()
print question3Output

Number of hidden units 1: 100, Number of hidden units 2: 100, Context window size: 7, Word embedding size: 50
verbose: True
nhidden2: 100
win: 7
nhidden1: 100
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 20
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 56.16 best test F1 51.84
NEW BEST: epoch 1 valid F1 65.37 best test F1 62.78
NEW BEST: epoch 2 valid F1 72.55 best test F1 70.22
NEW BEST: epoch 3 valid F1 77.2 best test F1 75.09
NEW BEST: epoch 4 valid F1 81.61 best test F1 79.0
NEW BEST: epoch 5 valid F1 83.25 best test F1 80.83
NEW BEST: epoch 6 valid F1 85.43 best test F1 82.55
NEW BEST: epoch 7 valid F1 86.94 best test F1 84.14
NEW BEST: epoch 8 valid F1 87.83 best test F1 85.2
NEW BEST: epoch 9 valid F1 88.88 best test F1 85.41
NEW BEST: epoch 10 valid F1 90.28 best test F1 87.14
NEW BEST: epoch 11 valid F1 90.31 best test F1 87.41
NEW BEST: epoch 12 valid F1 92.33 best test

### a.iii Implement RNN with two hidden layers
Testing accuracy accross 16 iterations is as follows:

```
[hidden units 1,hidden units 2, context window, embedding dimension -------- Testing accuracy]
  100,         , 100,             7,           50                            91.08
  100,         , 100,             7,           60                            91.12
  100,         , 100,            15,           50                            90.42
  100,         , 100,            15,           60                            90.9
  100,         , 200,             7,           50                            91.92
  100,         , 200,             7,           60                            91.9
  100,         , 200,            15,           50                            91.81
  100,         , 200,            15,           60                            91.63
  200,         , 100,             7,           50                            91.63
  200,         , 100,             7,           60                            92.54
  200,         , 100,            15,           50                            91.4
  200,         , 100,            15,           60                            91.59
  200,         , 200,             7,           50                            92.23
  200,         , 200,             7,           60                            92.25
  200,         , 200,            15,           50                            91.93
  200,         , 200,            15,           60                            91.13
```

The testing accuracy for RNN with 2 hidden layers is marginally better than the testing accuracy with one hidden layer. Also, we see that the testing accuracy increases as the total number of hidden units across two layers in RNN increase.

#### Bullet 4: Experiment with word embeddings

In [6]:
# TODO
from hw4a import test_rnnslu
import gc
gc.collect()
test_accuracy, emb, idx2word = test_rnnslu(fold=3, lr=0.0970806646812754, verbose=True, decay=True, win=7,
            nhidden=100, seed=345, emb_dimension=50, nepochs=5)
gc.collect()

verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 5
nhidden: 100
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 63.39 best test F1 60.51
NEW BEST: epoch 1 valid F1 75.46 best test F1 72.37
NEW BEST: epoch 2 valid F1 80.34 best test F1 77.06
NEW BEST: epoch 3 valid F1 83.89 best test F1 80.94
NEW BEST: epoch 4 valid F1 85.89 best test F1 82.83
BEST RESULT: epoch 4 valid F1 85.89 best test F1 82.83 with the model ../result


460695

In [7]:
from scipy import spatial
import gc
gc.collect()
embeddings = emb.get_value()
word_emb = {}
for idx, val in idx2word.iteritems():
    if "to" == val:
        word_emb[val] = embeddings[idx]
    if "UNK" not in val and "DIGIT" not in val and "'" not in val and len(val)>4:
        word_emb[val] = embeddings[idx]
print len(word_emb)
diff_dict = {}
for key1 in word_emb:
    for key2 in word_emb:
        rkey = key2 + '-' + key1
        if key1 is not key2 and rkey not in diff_dict:
            key = key1 + '-' + key2
            diff_dict[key] = word_emb[key1] - word_emb[key2]
print spatial.distance.cosine(diff_dict['united-reservation'], diff_dict['american-reservation'])
print spatial.distance.cosine(diff_dict['seattle-pittsburgh'], diff_dict['seattle-atlanta'])
print spatial.distance.cosine(diff_dict['oakland-delta'], diff_dict['oakland-continental'])
print spatial.distance.cosine(diff_dict['dallas-cities'], diff_dict['denver-cities'])
print spatial.distance.cosine(diff_dict['petersburg-atlanta'], diff_dict['petersburg-boston'])

357
0.0117974328026
0.0922372575811
0.0230399178163
0.0857083594518
0.0746536754777


### a.iv Experiment with word embeddings

After training the RNN model, we queried the embeddings and we zeroed down on the below mentioned relationships between word pairs:
```
F(united)-F(reservation) is similar to F(american)-F(reservation)
```
This shows that people were searching for reservations in two airlines - united and american

```
F(seattle)-F(pittsburgh) is similar to F(seattle)-F(atlanta)
```
This shows that people search for routes between seattle and pittsburgh as well as seattle and atlanta in the ATIS dataset sentences

```
F(oakland)-F(delta) is similar to F(oakland)-F(continental)
```
This shows that people were looking up delta and continental flights information to oakland

```
F(dallas)-F(cities) is similar to F(denver)-F(cities)
```
This shows that the model learnt that dallas and denver are cities

```
F(petersburg)-F(atlanta) is similar to F(petersburg)-F(boston)
```
This shows that people were looking for flights between petersburg and atlanta and petersburg and boston

### Problem b: Learning Parity Function
#### Bullet 1: Shallow MLP

In [1]:
# TODO
from hw4b import test_mlp_parity
import gc
gc.collect()
optimum_config_found = False
for lr in [1.0,0.9,0.8,0.7,0.6,0.5]:
    n_hidden = 8
    while n_hidden <= 128:
        print "Learning rate: %s, Hidden Units: %s"%(lr,n_hidden)
        test_acc = test_mlp_parity(learning_rate=lr, L1_reg=0.00, L2_reg=0.0000, n_epochs=100,
             batch_size=32, n_hidden=n_hidden, n_hiddenLayers=1,
             verbose=False,input_bit=8)
        n_hidden = n_hidden*2
        print test_acc
        gc.collect()
        if test_acc == 100.:
            print "Optimum configuration: Learning rate: %s, Hidden Units: %s"%(lr,n_hidden)
            optimum_config_found = True
            break
    if optimum_config_found:
        break

Learning rate: 1.0, Hidden Units: 8
... building the model
... training
Optimization complete.

Using gpu device 0: GRID K520
The training process for function test_mlp_parity ran for 0.04m


Best validation score of 1.458333 % obtained at iteration 2480, with test performance 1.041667 %
98.9583333333
Learning rate: 1.0, Hidden Units: 16
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 4.166667 % obtained at iteration 3069, with test performance 4.166667 %
95.8333333333
Learning rate: 1.0, Hidden Units: 32
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 53.125000 % obtained at iteration 217, with test performance 50.000000 %
50.0
Learning rate: 1.0, Hidden Units: 64
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.05m


Best validation score of 49.583333 % obtained at iteration 62, with test performance 45.833333 %
54.1666666667
Learning rate: 1.0, Hidden Units: 128
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 45.208333 % obtained at iteration 2852, with test performance 47.916667 %
52.0833333333
Learning rate: 0.9, Hidden Units: 8
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 33.541667 % obtained at iteration 3007, with test performance 29.166667 %
70.8333333333
Learning rate: 0.9, Hidden Units: 16
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 50.416667 % obtained at iteration 3069, with test performance 52.083333 %
47.9166666667
Learning rate: 0.9, Hidden Units: 32
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 49.166667 % obtained at iteration 496, with test performance 46.875000 %
53.125
Learning rate: 0.9, Hidden Units: 64
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 48.958333 % obtained at iteration 31, with test performance 47.916667 %
52.0833333333
Learning rate: 0.9, Hidden Units: 128
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 47.916667 % obtained at iteration 1798, with test performance 43.750000 %
56.25
Learning rate: 0.8, Hidden Units: 8
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 0.208333 % obtained at iteration 2604, with test performance 0.000000 %
100.0
Optimum configuration: Learning rate: 0.8, Hidden Units: 16


In [9]:
from hw4b import test_mlp_parity
from numpy import arange
import gc
gc.collect()
optimum_config_found = False
for i in arange(2.3, 1.4, -0.1):
    lr = float(i)
    n_hidden = 12
    while n_hidden <= 248832:
        print "Learning rate: %s, Hidden Units: %s"%(lr,n_hidden)
        test_acc = test_mlp_parity(learning_rate=lr, L1_reg=0.00, L2_reg=0.00, n_epochs=100,
             batch_size=32, n_hidden=n_hidden, n_hiddenLayers=1,
             verbose=False,input_bit=12)
        n_hidden = n_hidden*12
        print test_acc
        gc.collect()
        if test_acc == 100.:
            print "Optimum configuration: Learning rate: %s, Hidden Units: %s"%(lr,n_hidden)
            optimum_config_found = True
            break
    if optimum_config_found:
        break

Learning rate: 2.3, Hidden Units: 12
... building the model
... training
Optimization complete.
Best validation score of 45.833333 % obtained at iteration 3007, with test performance 54.166667 %
45.8333333333
Learning rate: 2.3, Hidden Units: 144
... building the model
... training
Optimization complete.

The training process for function test_mlp_parity ran for 0.04m
The training process for function test_mlp_parity ran for 0.04m


Best validation score of 16.041667 % obtained at iteration 2356, with test performance 12.500000 %
87.5
Learning rate: 2.3, Hidden Units: 1728
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 8.333333 % obtained at iteration 1891, with test performance 14.583333 %
85.4166666667
Learning rate: 2.3, Hidden Units: 20736
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.15m


Best validation score of 9.375000 % obtained at iteration 1643, with test performance 11.458333 %
88.5416666667
Learning rate: 2.3, Hidden Units: 248832
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 1.41m


Best validation score of 5.208333 % obtained at iteration 1643, with test performance 9.375000 %
90.625
Learning rate: 2.2, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 43.958333 % obtained at iteration 2790, with test performance 50.000000 %
50.0
Learning rate: 2.2, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 18.333333 % obtained at iteration 2077, with test performance 25.000000 %
75.0
Learning rate: 2.2, Hidden Units: 1728
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 10.625000 % obtained at iteration 1860, with test performance 7.291667 %
92.7083333333
Learning rate: 2.2, Hidden Units: 20736
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.15m


Best validation score of 8.750000 % obtained at iteration 1705, with test performance 3.125000 %
96.875
Learning rate: 2.2, Hidden Units: 248832
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 1.41m


Best validation score of 3.750000 % obtained at iteration 1643, with test performance 3.125000 %
96.875
Learning rate: 2.1, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 40.416667 % obtained at iteration 2883, with test performance 36.458333 %
63.5416666667
Learning rate: 2.1, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 15.000000 % obtained at iteration 2449, with test performance 25.000000 %
75.0
Learning rate: 2.1, Hidden Units: 1728
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 7.291667 % obtained at iteration 1984, with test performance 11.458333 %
88.5416666667
Learning rate: 2.1, Hidden Units: 20736
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.15m


Best validation score of 6.041667 % obtained at iteration 1643, with test performance 10.416667 %
89.5833333333
Learning rate: 2.1, Hidden Units: 248832
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 1.40m


Best validation score of 3.750000 % obtained at iteration 1612, with test performance 5.208333 %
94.7916666667
Learning rate: 2.0, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 16.666667 % obtained at iteration 3007, with test performance 18.750000 %
81.25
Learning rate: 2.0, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 11.041667 % obtained at iteration 3007, with test performance 14.583333 %
85.4166666667
Learning rate: 2.0, Hidden Units: 1728
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 8.750000 % obtained at iteration 1829, with test performance 10.416667 %
89.5833333333
Learning rate: 2.0, Hidden Units: 20736
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.15m


Best validation score of 7.291667 % obtained at iteration 1736, with test performance 12.500000 %
87.5
Learning rate: 2.0, Hidden Units: 248832
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 1.41m


Best validation score of 3.541667 % obtained at iteration 1674, with test performance 3.125000 %
96.875
Learning rate: 1.9, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 26.666667 % obtained at iteration 3069, with test performance 30.208333 %
69.7916666667
Learning rate: 1.9, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 14.583333 % obtained at iteration 2356, with test performance 17.708333 %
82.2916666667
Learning rate: 1.9, Hidden Units: 1728
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 8.541667 % obtained at iteration 1984, with test performance 9.375000 %
90.625
Learning rate: 1.9, Hidden Units: 20736
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.15m


Best validation score of 9.583333 % obtained at iteration 1736, with test performance 8.333333 %
91.6666666667
Learning rate: 1.9, Hidden Units: 248832
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 1.41m


Best validation score of 6.250000 % obtained at iteration 1829, with test performance 7.291667 %
92.7083333333
Learning rate: 1.8, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.03m


Best validation score of 47.083333 % obtained at iteration 62, with test performance 50.000000 %
50.0
Learning rate: 1.8, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 12.500000 % obtained at iteration 2325, with test performance 19.791667 %
80.2083333333
Learning rate: 1.8, Hidden Units: 1728
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 10.625000 % obtained at iteration 2108, with test performance 12.500000 %
87.5
Learning rate: 1.8, Hidden Units: 20736
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.15m


Best validation score of 6.041667 % obtained at iteration 2139, with test performance 5.208333 %
94.7916666667
Learning rate: 1.8, Hidden Units: 248832
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 1.41m


Best validation score of 9.791667 % obtained at iteration 1922, with test performance 10.416667 %
89.5833333333
Learning rate: 1.7, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 45.416667 % obtained at iteration 3100, with test performance 59.375000 %
40.625
Learning rate: 1.7, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 11.250000 % obtained at iteration 2604, with test performance 12.500000 %
87.5
Learning rate: 1.7, Hidden Units: 1728
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 8.125000 % obtained at iteration 2294, with test performance 15.625000 %
84.375
Learning rate: 1.7, Hidden Units: 20736
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.15m


Best validation score of 5.416667 % obtained at iteration 2201, with test performance 8.333333 %
91.6666666667
Learning rate: 1.7, Hidden Units: 248832
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 1.41m


Best validation score of 5.416667 % obtained at iteration 1798, with test performance 9.375000 %
90.625
Learning rate: 1.6, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 42.083333 % obtained at iteration 3100, with test performance 52.083333 %
47.9166666667
Learning rate: 1.6, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 19.375000 % obtained at iteration 3069, with test performance 11.458333 %
88.5416666667
Learning rate: 1.6, Hidden Units: 1728
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 10.833333 % obtained at iteration 2542, with test performance 11.458333 %
88.5416666667
Learning rate: 1.6, Hidden Units: 20736
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.15m


Best validation score of 6.875000 % obtained at iteration 2108, with test performance 7.291667 %
92.7083333333
Learning rate: 1.6, Hidden Units: 248832
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 1.41m


Best validation score of 5.000000 % obtained at iteration 2325, with test performance 6.250000 %
93.75
Learning rate: 1.5, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 47.500000 % obtained at iteration 279, with test performance 58.333333 %
41.6666666667
Learning rate: 1.5, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 17.708333 % obtained at iteration 2945, with test performance 19.791667 %
80.2083333333
Learning rate: 1.5, Hidden Units: 1728
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 11.041667 % obtained at iteration 2666, with test performance 16.666667 %
83.3333333333
Learning rate: 1.5, Hidden Units: 20736
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.15m


Best validation score of 5.208333 % obtained at iteration 2232, with test performance 5.208333 %
94.7916666667
Learning rate: 1.5, Hidden Units: 248832
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 1.41m


Best validation score of 4.375000 % obtained at iteration 2418, with test performance 7.291667 %
92.7083333333


### b.i Shallow MLP

I was able to achieve 100% accuracy for 8-bit inputs. The configuration used was Learning rate: 0.8, Hidden Units: 16

I wasn't able to achieve 100% accuracy for 12-bit inputs. The best accurcy achieved was 96.875% for the Learning rate: 2.2, Hidden Units: 20736

#### Bullet 2: Deep MLP

In [2]:
# TODO

from hw4b import test_mlp_parity
import gc
gc.collect()
optimum_config_found = False
for lr in [1.0,0.9,0.8,0.7,0.6,0.5]:
    n_hidden = 8
    while n_hidden <= 128:
        print "Learning rate: %s, Hidden Units: %s"%(lr,n_hidden)
        test_acc = test_mlp_parity(learning_rate=lr, L1_reg=0.00, L2_reg=0.0000, n_epochs=100,
             batch_size=32, n_hidden=n_hidden, n_hiddenLayers=2,
             verbose=False,input_bit=8)
        n_hidden = n_hidden*2
        print test_acc
        gc.collect()
        if test_acc == 100.:
            print "Optimum configuration: Learning rate: %s, Hidden Units: %s"%(lr,n_hidden)
            optimum_config_found = True
            break
    if optimum_config_found:
        break
gc.collect()
optimum_config_found = False
for i in arange(3, 1.5, -0.1):
    lr = float(i)
    n_hidden = 12
    while n_hidden <= 144:
        print "Learning rate: %s, Hidden Units: %s"%(lr,n_hidden)
        test_acc = test_mlp_parity(learning_rate=lr, L1_reg=0.00, L2_reg=0.0000, n_epochs=100,
             batch_size=32, n_hidden=n_hidden, n_hiddenLayers=2,
             verbose=False,input_bit=12)
        n_hidden = n_hidden*12
        print test_acc
        gc.collect()
        if test_acc == 100.:
            print "Optimum configuration: Learning rate: %s, Hidden Units: %s"%(lr,n_hidden)
            optimum_config_found = True
            break
    if optimum_config_found:
        break

Learning rate: 1.0, Hidden Units: 8
... building the model
... training
Optimization complete.
Best validation score of 2.708333 % obtained at iteration 3069, with test performance 3.125000 %
96.875
Learning rate: 1.0, Hidden Units: 16
... building the model
... training
Optimization complete.

The training process for function test_mlp_parity ran for 0.04m
The training process for function test_mlp_parity ran for 0.04m


Best validation score of 32.500000 % obtained at iteration 2542, with test performance 36.458333 %
63.5416666667
Learning rate: 1.0, Hidden Units: 32
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.04m


Best validation score of 0.833333 % obtained at iteration 2263, with test performance 0.000000 %
100.0
Optimum configuration: Learning rate: 1.0, Hidden Units: 64
Learning rate: 3.0, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 49.330357 % obtained at iteration 105, with test performance 50.000000 %
50.0
Learning rate: 3.0, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 47.767857 % obtained at iteration 15, with test performance 46.875000 %
53.125
Learning rate: 2.9, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 45.982143 % obtained at iteration 45, with test performance 45.312500 %
54.6875
Learning rate: 2.9, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 49.330357 % obtained at iteration 30, with test performance 48.437500 %
51.5625
Learning rate: 2.8, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 46.428571 % obtained at iteration 945, with test performance 37.500000 %
62.5
Learning rate: 2.8, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 49.553571 % obtained at iteration 15, with test performance 54.687500 %
45.3125
Learning rate: 2.7, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 49.776786 % obtained at iteration 15, with test performance 43.750000 %
56.25
Learning rate: 2.7, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 47.991071 % obtained at iteration 30, with test performance 46.875000 %
53.125
Learning rate: 2.6, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 44.642857 % obtained at iteration 15, with test performance 42.187500 %
57.8125
Learning rate: 2.6, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 49.776786 % obtained at iteration 15, with test performance 48.437500 %
51.5625
Learning rate: 2.5, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 49.330357 % obtained at iteration 30, with test performance 65.625000 %
34.375
Learning rate: 2.5, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 48.883929 % obtained at iteration 15, with test performance 43.750000 %
56.25
Learning rate: 2.4, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 46.651786 % obtained at iteration 1365, with test performance 48.437500 %
51.5625
Learning rate: 2.4, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 46.428571 % obtained at iteration 15, with test performance 43.750000 %
56.25
Learning rate: 2.3, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 48.214286 % obtained at iteration 15, with test performance 46.875000 %
53.125
Learning rate: 2.3, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 49.553571 % obtained at iteration 15, with test performance 32.812500 %
67.1875
Learning rate: 2.2, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 45.089286 % obtained at iteration 45, with test performance 59.375000 %
40.625
Learning rate: 2.2, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 49.553571 % obtained at iteration 15, with test performance 50.000000 %
50.0
Learning rate: 2.1, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 38.392857 % obtained at iteration 885, with test performance 46.875000 %
53.125
Learning rate: 2.1, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 49.553571 % obtained at iteration 15, with test performance 40.625000 %
59.375
Learning rate: 2.0, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 32.142857 % obtained at iteration 1080, with test performance 31.250000 %
68.75
Learning rate: 2.0, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 47.767857 % obtained at iteration 30, with test performance 62.500000 %
37.5
Learning rate: 1.9, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 45.535714 % obtained at iteration 660, with test performance 57.812500 %
42.1875
Learning rate: 1.9, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 45.535714 % obtained at iteration 855, with test performance 43.750000 %
56.25
Learning rate: 1.8, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 35.044643 % obtained at iteration 1485, with test performance 31.250000 %
68.75
Learning rate: 1.8, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.02m


Best validation score of 47.544643 % obtained at iteration 75, with test performance 64.062500 %
35.9375
Learning rate: 1.7, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.03m


Best validation score of 37.723214 % obtained at iteration 1380, with test performance 43.750000 %
56.25
Learning rate: 1.7, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.03m


Best validation score of 48.883929 % obtained at iteration 15, with test performance 51.562500 %
48.4375
Learning rate: 1.6, Hidden Units: 12
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.03m


Best validation score of 47.991071 % obtained at iteration 15, with test performance 40.625000 %
59.375
Learning rate: 1.6, Hidden Units: 144
... building the model
... training
Optimization complete.


The training process for function test_mlp_parity ran for 0.03m


Best validation score of 50.000000 % obtained at iteration 15, with test performance 54.687500 %
45.3125


### b.ii Deep MLP

I was able to achieve 100% accuracy for 8-bit inputs. The configuration used was Learning rate: 1.0, Hidden Units: 64 and 2 hidden layers

I wasn't able to achieve 100% accuracy for 12-bit input. The best accuracy achieved was 67.1875 for Learning rate: 2.3, Hidden Units: 144


#### Bullet 3: RNN

In [4]:
# TODO

from hw4b import test_rnn_parity
import gc
gc.collect()

test_acc = test_rnn_parity(learning_rate=1, L1_reg=0.00, L2_reg=0.0000, n_epochs=100,
             batch_size=32, n_hidden=64,verbose=False,input_bit=8)
print test_acc
gc.collect()

test_acc = test_rnn_parity(learning_rate=2.3, L1_reg=0.00, L2_reg=0.0000, n_epochs=100,
             batch_size=32, n_hidden=144,verbose=False,input_bit=12)
print test_acc

... building the model
... training
Optimization complete.
Best validation score of 44.791667 % obtained at iteration 31, with test performance 53.125000 %
46.875
... building the model
... training
Optimization complete.

The training process for function test_rnn_parity ran for 1.30m
The training process for function test_rnn_parity ran for 1.31m


Best validation score of 49.166667 % obtained at iteration 124, with test performance 42.708333 %
57.2916666667


### b.iii RNN

A RNN requires less number of neurons to achieve similar accuracy, however, this wasn't observed in the above experiment. One of the potential reasons for this could be that I used one layer in RNN. If we increase the number of neurons and hidden layers in RNN, we might be able to see increase in the testing accuracy.